Make `isSelfClosed` check more explicit

AndrewJakubowicz opened the pull request:
### Context

This was inspired (and maybe closes) https://github.com/runem/lit-analyzer/issues/283.

This may be the wrong direction for fixing the linked issue, however I wanted to propose a solution after my investigation.

### Current check

The current `no-unclosed-tag` check grabs the source code locations for the tag from `parse5`, and then tries to guess if the tag was closed.

The tag _is marked as self closing_ if the node has **no children** and the `startTag.endOffset === node.endOffset`.

An important note here, is that the source code locations are taken from `parse5`, which behaves similarly to the browser and thus has special parsing rules for various tags.

The edge case can be seen with `<p>` which can be closed implicitly.

```html
<p><div></div><!--becomes:--><p></p><div></div>
<p>Some content<div></div><!--becomes:--><p>Some content</p><div></div>
```

Currently, the first case without content will not raise the `no-unclosed-tag` check, but the second case does raise the `no-unclosed-tag` check.

Why? Because `parse5` correctly creates the DOM `<p>Some content</p><div></div>`, but then there is no reference to an authored closing tag, and the presence of child content fails to mark the tag as self closing.

### Possible change

This PR introduces a possible change where the intent of self closing can be marked by adding `/>`. Therefore implicit closing tags result in the `no-unclosed-tag` diagnostic being raised. To mark a tag as explicit, add the `/>` syntax (which FYI isn't a valid HTML syntax).

### Risk

This will catch more cases, so I'm not very stoked about this idea. I'm also not excited by the general concept of "self-closing" syntax which this PR pushes forward.

43081j commented:
maybe we can actually just check that there is an end tag?

for example:

```html
  <p>i never close
```

will have a `startTag` location, and no `endTag` location. 

similarly:

```html
  <p />
```

will behave the same since you can't self-close a `p` tag (parse5 will already be taking this into account i think). so it too won't have an `endTag` and will actually have a child text node (the whitespace following it).

so maybe we can just check `!getNodeSourceLocation(p5Node).endTag`?

i'm not sure why we didn't do that all along... unless i'm missing something 🤔

AndrewJakubowicz commented:
cc @43081j  and @rictic 

This is ready for review. Thank you James for your exceptional feedback!

43081j commented:
the CI failure is because one of the tests actually self-closes a custom element FYI 😬 

so it already caught one!

43081j reviewed (approved):
looks good to me 👍 

we should probably double check if it does solve the issue you mentioned too. that was more about a space tripping it up i think

AndrewJakubowicz commented:
Excellent! :D

1. Added a specific test case for `<p> <div></div></p>` (which continues to raise the diagnostic).
2. Fixed the self-closing custom element in the other test.

rictic commented on packages/lit-analyzer/src/lib/rules/no-unclosed-tag.ts:
```diff
@@ -11,18 +34,18 @@ const rule: RuleModule = {
-				location: rangeFromHtmlNode(htmlNode),
-				message: `This tag isn't closed.${isCustomElement ? " Custom elements cannot be self closing." : ""}`
-			});
+		if (VOID_ELEMENTS.has(htmlNode.tagName.toLowerCase()) || htmlNode.location.endTag != null) {
```
Do we warn on an invalid closing element for a void element? Seems like it'd be good to do so. For example, this should be a warning on the `</input>`

```html
<input><div></div></input>
```

AndrewJakubowicz replied:
I think this is a good idea but maybe something we can add incrementally to the `no-unclosed-tag` diagnostic.

I'll make an issue.

rictic commented on packages/lit-analyzer/src/test/rules/no-unclosed-tag.ts:
```diff
@@ -7,7 +7,49 @@ tsTest("Report unclosed tags", t => {
 	hasDiagnostic(t, diagnostics, "no-unclosed-tag");
 });
 
```
In an ideal world, we might want to allow:

```html
<p>One
<p>Two
```

Or just:

```html
<p>Hi
```

And we'd want to put warnings on dangling closing tags that don't match up to any opening tags.

All of those may be somewhat tricky to do with how parse5 represents the tree though, requiring us to notice spans in the original text that aren't represented in the parsed tree.

AndrewJakubowicz replied:
These are very tricky, and another good case to add to an issue for followup.

Philosophically though, I wonder if the lit-analyzer should avoid HTML diagnostics and leave that to an HTML specific tool. It may even be worth scoping the `no-unclosed-tag` to only check custom element tags.
